docs/settings.rst
.. _settings:

Settings
========

Settings reference
~~~~~~~~~~~~~~~~~~


Here is a complete list of settings which control default PynamoDB behavior.

connect_timeout_seconds
-----------------------

Default: ``15``

The time in seconds till a ``ConnectTimeoutError`` is thrown when attempting to make a connection.


read_timeout_seconds
-----------------------

Default: ``30``

The time in seconds till a ``ReadTimeoutError`` is thrown when attempting to read from a connection.


max_retry_attempts
------------------

Default: ``3``

The number of times to retry certain failed DynamoDB API calls. The most common cases eligible for
retries include ``ProvisionedThroughputExceededException`` and ``5xx`` errors.


region
------

Default: ``"us-east-1"``

The default AWS region to connect to.


max_pool_connections
--------------------

Default: ``10``

The maximum number of connections to keep in a connection pool.


extra_headers
--------------------

Default: ``None``

A dictionary of headers that should be added to every request. This is only useful
when interfacing with DynamoDB through a proxy, where headers are stripped by the
proxy before forwarding along. Failure to strip these headers before sending to AWS
will result in an ``InvalidSignatureException`` due to request signing.


host
------

Default: automatically constructed by boto to account for region

The URL endpoint for DynamoDB. This can be used to use a local implementation of DynamoDB such as DynamoDB Local or dynalite.


retry_configuration
-------------------

Default: ``"LEGACY"``

This controls the PynamoDB retry behavior. The default of ``"LEGACY"`` keeps the
existing PynamoDB retry behavior. If set to ``None``, this will use botocore's default
retry configuration discovery mechanism as documented
`in boto3 <https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html#retries>`_
and
`in the AWS SDK docs <https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html>`_.
If set to a retry configuration dictionary as described
`here <https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html#defining-a-retry-configuration-in-a-config-object-for-your-boto3-client>`_
it will be used directly in the botocore client configuration.

Overriding settings
~~~~~~~~~~~~~~~~~~~

Default settings may be overridden by providing a Python module which exports the desired new values.
Set the ``PYNAMODB_CONFIG`` environment variable to an absolute path to this module or write it to
``/etc/pynamodb/global_default_settings.py`` to have it automatically discovered.



docs/release_notes.rst
.. _release_notes:

Release Notes
=============

v6.1.0
------

Features:

* Add the ability to set or unset the boto retry configuration (:pr:`1271`)

    * This adds the ability to directly set the boto retry configuration dictionary, or
      to leave it unset and allow botocore to automatically discover the configuration
      from the environment or `~/.aws/config` files.
    * A new setting `retry_configuration` was added to configure this behavior.
    * The default behavior of PynamoDB in regards to configuring retries remains the
      same as before.

v6.0.2
------

Fixes:

* Fix a warning about `datetime.utcfromtimestamp` deprecation (:pr:`1261`)

v6.0.1
------

Features:

* For failed transaction, return the underlying item in :code:`cancellation_reasons[...].raw_item` (:pr:`1226`).
  This only applies when passing :code:`return_values=ALL_OLD`.

Fixes:

* Fixing (:pr:`1242`) regression to the :code:`extra_headers` feature. These headers are intended for proxies that strip them,
  so they should be excluded from the AWS signature.

v6.0.0
------

This is a major release and contains breaking changes. Please read the notes below carefully.

Breaking changes:

* :py:class:`~pynamodb.attributes.BinaryAttribute` and :py:class:`~pynamodb.attributes.BinarySetAttribute` have undergone breaking changes:

  * The attributes' internal encoding has changed. To prevent this change going unnoticed, a new required :code:`legacy_encoding` parameter was added: see :doc:`upgrading_binary` for details.
    If your codebase uses :py:class:`~pynamodb.attributes.BinaryAttribute` or :py:class:`~pynamodb.attributes.BinarySetAttribute`,
    go over the attribute declarations and mark them accordingly.
  * When using binary attributes, the return value of :meth:`~pynamodb.models.Model.serialize` will no longer be JSON-serializable
    since it will contain :code:`bytes` objects. Use :meth:`~pynamodb.attributes.AttributeContainer.to_dynamodb_dict` and :meth:`~pynamodb.attributes.AttributeContainer.to_simple_dict` for JSON-serializable mappings.
    for a safe JSON-serializable representation.

* Python 3.6 is no longer supported.
* PynamoDB no longer has a default AWS region (used to be us-east-1) (:pr:`1003`).
  If needed, update your models' `Meta` or set the `AWS_DEFAULT_REGION` environment variable.
* :py:class:`~pynamodb.models.Model`'s JSON serialization helpers were changed:

  * :code:`to_json` was renamed to :meth:`~pynamodb.attributes.AttributeContainer.to_simple_dict` (:pr:`1126`). Additionally, :meth:`~pynamodb.attributes.AttributeContainer.to_dynamodb_dict`
    and :meth:`~pynamodb.attributes.AttributeContainer.from_dynamodb_dict` were added for round-trip JSON serialization.
  * :code:`pynamodb.util.attribute_value_to_json` was removed (:pr:`1126`)

* :py:class:`~pynamodb.attributes.Attribute`'s :code:`default` parameter must be either an immutable value
  (of one of the built-in immutable types) or a callable.
  This prevents a common class of errors caused by unintentionally mutating the default value.
  A simple workaround is to pass an initializer (e.g. change :code:`default={}` to
  :code:`default=dict`) or wrap in a lambda (e.g. change :code:`default={'foo': 'bar'}` to
  :code:`default=lambda: {'foo': 'bar'}`).
* :meth:`~pynamodb.indexes.Index.count`, :meth:`~pynamodb.indexes.Index.query`,
  and :meth:`~pynamodb.indexes.Index.scan` are now instance methods.
* :py:class:`~pynamodb.settings.OperationSettings` has been removed.

Major changes:

* We are now compatible with `opentelemetry botocore instrumentation <https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-botocore>`_.
* We've reduced our usage of botocore private APIs (:pr:`1079`). On multiple occasions, new versions
  of botocore broke PynamoDB, and this change lessens the likelihood of that happening in the future
  by reducing (albeit not eliminating) our reliance on private botocore APIs.

Minor changes:

* :meth:`~pynamodb.models.Model.save`, :meth:`~pynamodb.models.Model.update`, :meth:`~pynamodb.models.Model.delete_item`,
  and :meth:`~pynamodb.models.Model.delete` now accept a ``add_version_condition`` parameter.
  See :ref:`optimistic_locking_version_condition` for more details.
* :meth:`~pynamodb.models.Model.batch_get`, has guard rails defending against items without a hash_key and range_key.
* :meth:`~pynamodb.attributes.Attribute.set`, can remove attribute by assigning an empty value in the update expression.

v5.5.1
----------
* Fix compatibility with botocore 1.33.2 (#1205)

v5.5.0
----------
* :meth:`~pynamodb.models.Model.save`, :meth:`~pynamodb.models.Model.update`, :meth:`~pynamodb.models.Model.delete_item`,
  and :meth:`~pynamodb.models.Model.delete` now accept a ``add_version_condition`` parameter.
  See :ref:`optimistic_locking_version_condition` for more details.

v5.4.1
----------
* Use model's AWS credentials in threads (#1164)

  A model can specify custom AWS credentials in the ``Meta`` class (in lieu of "global"
  AWS credentials from the environment). Previously those model-specific credentials
  were not used from within new threads.

Contributors to this release:

* @atsuoishimoto

v5.4.0
----------
* Expose transaction cancellation reasons in
  :meth:`~pynamodb.exceptions.TransactWriteError.cancellation_reasons` and
  :meth:`~pynamodb.exceptions.TransactGetError.cancellation_reasons` (#1144).

v5.3.2
----------
* Prevent ``typing_tests`` from being installed into site-packages (:pr:`1118`)

Contributors to this release:

* :user:`musicinmybrain`


v5.3.1
----------
* Fixed issue introduced in 5.3.0: using :py:class:`~pynamodb.connection.table.TableConnection` directly (not through a model)
  raised the following exception::

    pynamodb.exceptions.TableError: Meta-table for '(table-name)' not initialized

* Fix typing on :py:class:`~pynamodb.transactions.TransactGet` (backport of :pr:`1057`)


v5.3.0
----------

pynamodb/settings.py
import importlib.util
import logging
import os
import warnings
from os import getenv

from typing import Any

log = logging.getLogger(__name__)

default_settings_dict = {
    'connect_timeout_seconds': 15,
    'read_timeout_seconds': 30,
    'max_retry_attempts': 3,
    'region': None,
    'max_pool_connections': 10,
    'extra_headers': None,
    'retry_configuration': 'LEGACY'
}

OVERRIDE_SETTINGS_PATH = getenv('PYNAMODB_CONFIG', '/etc/pynamodb/global_default_settings.py')


def _load_module(name, path):
    # https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)  # type: ignore
    spec.loader.exec_module(module)  # type: ignore
    return module


override_settings = {}
if os.path.isfile(OVERRIDE_SETTINGS_PATH):
    override_settings = _load_module('__pynamodb_override_settings__', OVERRIDE_SETTINGS_PATH)
    if hasattr(override_settings, 'session_cls') or hasattr(override_settings, 'request_timeout_seconds'):
        warnings.warn("The `session_cls` and `request_timeout_second` options are no longer supported")
    log.info('Override settings for pynamo available {}'.format(OVERRIDE_SETTINGS_PATH))
else:
    log.info('Override settings for pynamo not available {}'.format(OVERRIDE_SETTINGS_PATH))
    log.info('Using Default settings value')


def get_settings_value(key: str) -> Any:
    """
    Fetches the value from the override file.
    If the value is not present, then tries to fetch the values from constants.py
    """
    if hasattr(override_settings, key):
        return getattr(override_settings, key)

    if key in default_settings_dict:
        return default_settings_dict[key]

    return None
